File: fixtures/coding/maintainability-mini-service-python/service/validation.py
"""Validation helpers for customer payloads."""

from __future__ import annotations

from typing import Any

ALLOWED_PLANS = {"free", "pro", "enterprise"}


def validate_customer_payload(payload: dict[str, Any]) -> list[str]:
    """Validate the existing single-customer create payload."""
    errors: list[str] = []
    name = payload.get("name")
    email = payload.get("email")
    plan = payload.get("plan", "free")
    tags = payload.get("tags", [])

    if not isinstance(name, str) or not name.strip():
        errors.append("name is required")
    if not isinstance(email, str) or "@" not in email or "." not in email.split("@")[-1]:
        errors.append("email must be valid")
    if plan not in ALLOWED_PLANS:
        errors.append("plan is invalid")
    if not isinstance(tags, list) or not all(isinstance(tag, str) for tag in tags):
        errors.append("tags must be a list of strings")

    return errors


def normalize_customer_payload(payload: dict[str, Any]) -> dict[str, Any]:
    """Normalize the existing single-customer create payload."""
    return {
        "name": str(payload.get("name", "")).strip(),
        "email": str(payload.get("email", "")).strip().lower(),
        "plan": payload.get("plan", "free"),
        "tags": tuple(str(tag).strip().lower() for tag in payload.get("tags", [])),
    }

File: fixtures/coding/maintainability-mini-service-ts/src/validation.ts
// Validation helpers for customer payloads.

import type { Plan } from "./models";

const ALLOWED_PLANS = new Set<Plan>(["free", "pro", "enterprise"]);

export interface NormalizedCustomerPayload {
  name: string;
  email: string;
  plan: Plan;
  tags: readonly string[];
}

export function validateCustomerPayload(payload: Record<string, unknown>): string[] {
  const errors: string[] = [];
  const name = payload.name;
  const email = payload.email;
  const plan = payload.plan ?? "free";
  const tags = payload.tags ?? [];

  if (typeof name !== "string" || name.trim() === "") {
    errors.push("name is required");
  }
  if (typeof email !== "string" || !email.includes("@") || !email.split("@").at(-1)?.includes(".")) {
    errors.push("email must be valid");
  }
  if (typeof plan !== "string" || !ALLOWED_PLANS.has(plan as Plan)) {
    errors.push("plan is invalid");
  }
  if (!Array.isArray(tags) || !tags.every((tag) => typeof tag === "string")) {
    errors.push("tags must be a list of strings");
  }

  return errors;
}

export function normalizeCustomerPayload(payload: Record<string, unknown>): NormalizedCustomerPayload {
  const tags = Array.isArray(payload.tags) ? payload.tags : [];
  return {
    name: String(payload.name ?? "").trim(),
    email: String(payload.email ?? "").trim().toLowerCase(),
    plan: (payload.plan ?? "free") as Plan,
    tags: tags.map((tag) => String(tag).trim().toLowerCase()),
  };
}
